core/src/com/google/inject/internal/InternalContext.java

```diff
@@ -106,10 +98,36 @@ public void popState() {
     return builder.build();
   }
 
-  private static final class DependencyStack extends ArrayList<Object> {
-    void pop() {
-      int sz = size();
-      removeRange(sz - 2, sz);
+  /**
+   * Keeps track of the hierarchy of types needed during injection.
+   *
+   * <p>This is a pairwise combination of dependencies and sources, with dependencies or keys on
+   * even indices, and sources on odd indices. This structure is to avoid the memory overhead of
+   * DependencyAndSource objects, which can add to several tens of megabytes in large applications.
+   */
+  private static final class DependencyStack {
+    private Object[] elements = new Object[10];
```

Comment:
I'm really sure what the usage patterns of `InternalContext` are like, but could it be worth delaying the allocation until the first dependency is pushed?